00X-ABI-Stable-Module-API.md

```diff
@@ -0,0 +1,369 @@
+  or further APIs provided by V8
+* Allows native modules built as shared libraries to be used
+  with different versions of Node.js
+* Makes it clear which APIs can use with the expectation
+  of not being affected by changes in the JavaScript engine
+  or Node.js release
+
+# Background
+
+There have been ongoing discussions on a new module API in both the
+community api working group (https://github.com/nodejs/api), other
+issues in the Node.js  issue tracker, and the recent community Node.js
+face-to-face VM summit (https://github.com/jasnell/vm-summit).
+
+In that summit there was agreement that we wanted to define this
+module API.  The general approach was to be:
+
+* Start with NAN, trim down to minimal subset based on module usage
+  and inspection.
+* Add in VM agnostic methods for object lifecycle management.
+* Add in VM agnostic exception management
+* Define C/C++ types for the API which are independent from V8
+  (NAN currently uses the V8 types).
+* Use the NAN examples to work through the API and build up the
+  definition of the key methods.
+* Validate on most used modules in the ecosystem
+* Complete performance analysis to confirm accepted level of overhead.
+
+Notes/details on other related work and the current experimental progress
+on implementing this eps can be found in:
+https://docs.google.com/document/d/1bX9SZKufbfhr7ncXL9fFev2LrljymtFZdWpwmv7mB9g/edit#heading=h.ln9aj9qjo969
+
+# API definition
+
+**WORK IN PROGRESS, DRIVEN BY MODULES PORTED SO FAR**
+
+This API will be built up from a minimal set. Based on this set
+we have been able to port the NaN example and level down.  Additions
+will be made as we port additional modules.
+
+### Types
+```C
+typedef struct napi_env__ *napi_env;
+typedef struct napi_value__ *napi_value;
+typedef struct napi_persistent__ *napi_persistent;
+typedef struct napi_handle_scope__ *napi_handle_scope;
+typedef struct napi_escapable_handle_scope__ *napi_escapable_handle_scope;
+typedef struct napi_propertyname__ *napi_propertyname;
+typedef const struct napi_func_cb_info__ *napi_func_cb_info;
+typedef void (*napi_callback)(napi_env, napi_func_cb_info);
+typedef void napi_destruct(void* v);
+
+
+struct napi_method_descriptor {
+  napi_callback callback;
+  char* utf8name;
+};
+
+extern "C" {
+enum napi_valuetype {
+    // ES6 types (corresponds to typeof)
+    napi_undefined,
+    napi_null,
+    napi_boolean,
+    napi_number,
+    napi_string,
+    napi_symbol,
+    napi_object,
+    napi_function,
+};
+```
+### Functions
+
+```C
+// Getters for defined singletons
+NODE_EXTERN napi_value napi_get_undefined_(napi_env e);
+NODE_EXTERN napi_value napi_get_null(napi_env e);
+NODE_EXTERN napi_value napi_get_false(napi_env e);
+NODE_EXTERN napi_value napi_get_true(napi_env e);
+NODE_EXTERN napi_value napi_get_global_scope(napi_env e);
+
+
+// Methods to create Primitive types/Objects
+NODE_EXTERN napi_value napi_create_object(napi_env e);
```

Comment:
The goal at this point is to land the ep as a draft.  We expect there will be changes to the API's as we build up experience in using them to port/performance etc.  Once its in as a draft them we can have PR's that update based on the experimentation or even just suggestions.  So its not that suggestions aren't valid but that we are not trying to achieve the final shape in the DRAFT state.    Even better would be PR's to the code itself.